server/src/main/java/org/opensearch/gateway/AsyncBatchShardFetch.java

```diff
@@ -0,0 +1,488 @@
+        if (nodesToFetch.isEmpty() == false) {
+            // mark all node as fetching and go ahead and async fetch them
+            // use a unique round id to detect stale responses in processAsyncFetch
+            final long fetchingRound = round.incrementAndGet();
+            for (AsyncBatchShardFetch.NodeEntry<T> nodeEntry : nodesToFetch) {
+                nodeEntry.markAsFetching(fetchingRound);
+            }
+            DiscoveryNode[] discoNodesToFetch = nodesToFetch.stream()
+                .map(AsyncBatchShardFetch.NodeEntry::getNodeId)
+                .map(nodes::get)
+                .toArray(DiscoveryNode[]::new);
+            asyncFetch(discoNodesToFetch, fetchingRound);
+        }
+
+        // if we are still fetching, return null to indicate it
+        if (hasAnyNodeFetching(cache)) {
+            return new AsyncBatchShardFetch.FetchResult<>(null, emptyMap());
+        } else {
+            // nothing to fetch, yay, build the return value
+            Map<DiscoveryNode, T> fetchData = new HashMap<>();
+            Set<String> failedNodes = new HashSet<>();
+            for (Iterator<Map.Entry<String, AsyncBatchShardFetch.NodeEntry<T>>> it = cache.entrySet().iterator(); it.hasNext();) {
+                Map.Entry<String, AsyncBatchShardFetch.NodeEntry<T>> entry = it.next();
+                String nodeId = entry.getKey();
+                AsyncBatchShardFetch.NodeEntry<T> nodeEntry = entry.getValue();
+
+                DiscoveryNode node = nodes.get(nodeId);
+                if (node != null) {
+                    if (nodeEntry.isFailed()) {
+                        // if its failed, remove it from the list of nodes, so if this run doesn't work
+                        // we try again next round to fetch it again
+                        it.remove();
+                        failedNodes.add(nodeEntry.getNodeId());
+                    } else {
+                        if (nodeEntry.getValue() != null) {
+                            fetchData.put(node, nodeEntry.getValue());
+                        }
+                    }
+                }
+            }
+            Set<String> allIgnoreNodes = unmodifiableSet(new HashSet<>(nodesToIgnore));
+            // clear the nodes to ignore, we had a successful run in fetching everything we can
+            // we need to try them if another full run is needed
+            nodesToIgnore.clear();
+
+            // if at least one node failed, make sure to have a protective reroute
+            // here, just case this round won't find anything, and we need to retry fetching data
+            if (failedNodes.isEmpty() == false || allIgnoreNodes.isEmpty() == false) {
+                reroute(batchUUID, "nodes failed [" + failedNodes.size() + "], ignored [" + allIgnoreNodes.size() + "]");
+            }
+
+            return new AsyncBatchShardFetch.FetchResult<>(fetchData, ignoredShardToNodes);
```

Comment:
`ignoredShardToNodes` - this variable here will be empty in the first fetch right ?

Reply:
Not sure, RoutingNodes maintain the state, whatever it sends we store here as per previous behaviour

Reply:
I meant, you may want to send actual ignore nodes data i.e. **ignoreNodes** argument ?

ignoredShardToNodes - this is never updated, just an empty map according to this PR's code.